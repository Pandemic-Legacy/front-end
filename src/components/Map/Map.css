.Map {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  
  .svgStyle {
    background-color: none;
    display: block;
    width: 100%;
    height: 40vh;
    overflow: visible;
    circle:hover {
      cursor: grab;
    }
    path {
      stroke: rgba(159, 142, 160, 0.1);
    }
    path:hover {
      stroke: rgba(0, 0, 0, 0.3);
      stroke-width: 0.6px;
      cursor: pointer;
    }
    circle:active {
      cursor: grabbing;
    }
    path:active {
      cursor: grabbing;
    }
  }
  

}

.mapLegend {
  color: seashell;
  text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.5);
  font-size: 1rem;
  padding: 0.3rem;
}
.mapLegendContainer {
  display: flex;
  flex-direction: column;
  justify-content: stretch;
}

.legendNoData {
  color: seashell;
  text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.5);
  font-size: 1rem;
  padding: 0.3rem;
  margin: 0;
  background-color: silver;
}