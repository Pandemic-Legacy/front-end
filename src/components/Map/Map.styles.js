import { makeStyles } from '@material-ui/core';

export const useStyles = makeStyles((theme) => ({
  mapContainer: {
    minHeight: '600px',
    padding: '2rem',
    marginBottom: '1rem'
  },
  popover: {
    pointerEvents: 'auto',
    background: 'rgba(192, 199, 207, 0.7)'
  },
  paper: {
    padding: theme.spacing(3),
    textAlign: 'center',
    background: 'rgba(255, 255, 255, 0.8)'
  },
  legendPaper: {
    padding: theme.spacing(1),
    textAlign: 'center',
    boxShadow: '0 0 0'
  },
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
  arrow: {
    width: '3rem',
    zIndex: 3
  },
  dragLabel: {
    zIndex: 5,
    backgroundColor: 'rgba(240, 240, 240, 0.7)',
    padding: '1rem',
    borderRadius: '0.5rem',
    position: 'absolute'
  }
}));
