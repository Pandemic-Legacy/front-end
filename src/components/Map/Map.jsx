import React, { useEffect, useRef, useState } from 'react';
import { select, geoPath, geoOrthographic, scaleLinear, event, drag, geoMercator } from 'd3';
import { useResizeObserver } from '../../hooks/d3Hooks';
import PropTypes from 'prop-types';

import { Slider, Popover, Typography, Button, withStyles, FormControl, InputLabel, Select, MenuItem, Paper, Grid, FormLabel, RadioGroup, FormControlLabel, Radio } from '@material-ui/core'; 

import style from './Map.css';

import { useDispatch, useSelector } from 'react-redux';
import { setGlobalMobilityDataByDate, setSelectedCountryCode, setSelectedCountry } from '../../actions/actions';
import { getMobilityDates, getSelectedCountryCode } from '../../selectors/selectors';
import { useHistory } from 'react-router-dom';
import { useStyles } from './Map.styles';
// import { useIsMobile } from '../hooks/isMobile';

const SliderStyled = withStyles({
  root: {
    height: 8
  },
  thumb: {
    height: 20,
    width: 20,
    backgroundColor: '#fff',
    border: '2px solid currentColor',
    marginTop: -6,
    marginLeft: -12
  },
  active: {},
  valueLabel: {
    left: 'calc(-50%)'
  },
  track: {
    height: 8,
    borderRadius: 4
  },
  rail: {
    height: 8,
    borderRadius: 4,
    width: '100.5%'
  },
  mark: {
    height: 8,
    width: 2,
    marginTop: 0
  },
})(Slider);

const Map = ({ mapData, countryCode = '' }) => {
  const dates = useSelector(getMobilityDates);
  const selectedCountryCode =  useSelector(getSelectedCountryCode);

  const [property, setProperty] = useState('residentialChange');
  const [rotateX, setRotateX] = useState(0);
  const [rotateY, setRotateY] = useState(0);
  const [rotating, setRotating] = useState(false);
  const [dateIndex, setDateIndex] = useState(84); //hard coded index for now, would come from dates.length - 1
  const [selectedCountryName, setSelectedCountryName] = useState('test'); 
  const [selectedCountryData, setSelectedCountryData] = useState({});

  const classes = useStyles();

  //use increments of 21 for more even markers
  const marks = [
    { value: 0, label: dates[0]?.slice(5) },
    { value: 15, label: dates[15]?.slice(5) },
    { value: 29, label: dates[29]?.slice(5) },
    { value: 46, label: dates[46]?.slice(5) },
    { value: 60, label: dates[60]?.slice(5) },
    { value: 76, label: dates[76]?.slice(5) },
    { value: 84, label: dates[84]?.slice(5) },
  ];

  //PopOver
  const [anchorEl, setAnchorEl] = useState(null);
  // const handlePopoverOpen = (event) => {
  //   setAnchorEl(event.currentTarget);
  // };
  const handlePopoverClose = () => {
    setAnchorEl(null);
  };
  const open = Boolean(anchorEl);
  //PopOver

  const svgRef = useRef();
  const wrapperRef = useRef();
  const legendRef = useRef();
  const dimensions = useResizeObserver(wrapperRef);
  const wrapperHeight = dimensions?.height;
  const dispatch = useDispatch();
  const history = useHistory();
  // const isMobile = useIsMobile();
  

  //this could be trimed down if we used redux for countryName
  useEffect(() => {
    if(!selectedCountryCode) return;
    const countryData = mapData.features.find(country => country.mobilityData.countryCode === selectedCountryCode).mobilityData;
    setSelectedCountryName(countryData.countryName);
    setAnchorEl(wrapperRef.current);
  }, [selectedCountryCode]);

  useEffect(() => {
    if(!dates.length) return;
    else dispatch(setGlobalMobilityDataByDate(dates[dateIndex]));
  }, [dateIndex]);
  
  useEffect(() => {
    if(!mapData.features) return;

    const svg = select(svgRef.current);

    const { width, height } = dimensions || wrapperRef.current.getBoundingClientRect();
    
    const colorScale = scaleLinear()
      .domain([-100, 0, 100])
      .range(['#B71C1C', 'rgb(243, 240, 225)', '#1d7d0a']);

    const globePosition = [width / 2, height / 2];
    const globeScale = 1;
    //globe Projection
    const projection = geoOrthographic()
      .fitSize([width * globeScale, height * globeScale], mapData)
      .center([0, 0])
      .rotate([rotateX, rotateY, 0])
      .translate(globePosition)
      .precision(100);

    
    
    const selectedCountry = mapData.features
      .find(({ properties }) => properties.iso_a2 === countryCode);
    //Country Projection
    const flatProjection = geoMercator()
      .fitSize([width, height], selectedCountry)
      .precision(50);

    const pathGenerator = countryCode ? geoPath().projection(flatProjection) : geoPath().projection(projection);

    const defs = svg.append('defs');
    const linearGradient = defs.append('linearGradient')
      .attr('id', 'linear-gradient')
      .attr('x1', '60%')
      .attr('y1', '30%')
      .attr('x2', '20%')
      .attr('y2', '90%');
    linearGradient.append('stop')
      .attr('offset', '0%')
      .attr('stop-color', '#2493C3');
    linearGradient.append('stop')
      .attr('offset', '100%')
      .attr('stop-color', '#2C4099');

    const drop_shadow = svg.append('defs').append('radialGradient')
      .attr('id', 'drop_shadow')
      .attr('cx', '50%')
      .attr('cy', '50%');
    drop_shadow.append('stop')
      .attr('offset', '20%').attr('stop-color', '#000')
      .attr('stop-opacity', '.5');
    drop_shadow.append('stop')
      .attr('offset', '100%').attr('stop-color', '#000')
      .attr('stop-opacity', '0');  

    if(!countryCode) svg
      .selectAll('ellipse')
      .data(['spot'])
      .join('ellipse')
      .attr('cx', globePosition[0] - (width / 20))
      .attr('cy', globePosition[1] + (height / 2.06))
      .attr('rx', projection.scale() * 1.1)
      .attr('ry', projection.scale() * .25)
      .attr('class', 'noclicks')
      .style('fill', 'url(#drop_shadow)');

    if(!countryCode) svg
      .selectAll('circle')
      .data(['spot'])
      .join('circle')
      .attr('cx', globePosition[0])
      .attr('cy', globePosition[1])
      .attr('r', projection.scale())
      .style('fill', 'url(#linear-gradient)');


    
    if(!countryCode) svg.call(drag()
      .on('start', () => { setRotating(true);})
      .on('drag', () => {

        const rotate = projection.rotate();
        const sensitivity = 50 / projection.scale();

        projection.rotate([
          rotate[0] + event.dx * sensitivity, 
          rotate[1] - event.dy * sensitivity,
        ]);
        setRotateX(rotate[0] + event.dx * sensitivity);
        setRotateY(rotate[1] - event.dy * sensitivity);

        const path = geoPath().projection(projection);
        svg.selectAll('path').attr('d', path);
      })
      .on('end', () => { setRotating(false);})
      
    );

    const map = svg
      .selectAll('.country')
      .data(mapData.features)
      .join('path')
      .on('click', (country) => {
        const { countryCode, countryName } = country.mobilityData;
        dispatch(setSelectedCountry({ countryCode, countryName }));
        setSelectedCountryData(country.mobilityData);
      })
      .attr('class', 'country');
    
    if(rotating) {
      map
        .attr('fill', country => country.mobilityData[property] 
          ? colorScale(country.mobilityData[property])
          : '#dfe2e8'
        )
        .attr('d', country => pathGenerator(country));  
    } else {
      map
        .transition()
        .attr('fill', country => country.mobilityData[property] 
          ? colorScale(country.mobilityData[property])
          : '#dfe2e8'
        )
        .attr('d', country => pathGenerator(country));
    }
    
    const legend = select(legendRef.current);
    const legendText = [100, 75, 50, 25, 0, -25, -50, -75, -100];
    legend.selectAll('span')
      .data(legendText)
      .join('span')     
      .attr('class', style.mapLegend)
      .style('background', (d) => colorScale(d))
      .text((d, i) => legendText[i]);
      
  }, [mapData, dimensions, property]);

  return (<>
    <Grid container className={classes.mapContainer} alignItems="center" justify="center" spacing={2}>

      <Grid item xs={3} sm={2} >
        <Paper elevation={2} className={classes.legendPaper}>
          <div ref={legendRef} 
            className={style.mapLegendContainer}
          >Change in {property.replace('Change', '')} Mobility 
          </div>
          <p className={style.legendNoData}>No Data Available</p>
        </Paper>
      </Grid>
    
      <Grid item xs={9} sm={8}ref={wrapperRef} className={style.Map} >
        <svg ref={svgRef} className={style.svgStyle}></svg>
        <Popover id={style.countryPopover} 
          className={classes.popover} 
          classes={{ paper: classes.paper }} 
          open={open} 
          anchorEl={anchorEl} 
          anchorOrigin={{ vertical: 'center', horizontal: 'center' }}
          transformOrigin={{ vertical: wrapperHeight / 5, horizontal: 'center' }}
          onClose={handlePopoverClose}
          disableRestoreFocus
        >
          <Typography variant="h4">{selectedCountryName}</Typography>
          <Typography>
            {property.replace('Change', '')}: {selectedCountryData[property] || 'N/A'}
          </Typography>
          <Button variant="contained" 
            color="primary" 
            onClick={(e) => {
              e.preventDefault();
              history.push(`/country/${selectedCountryCode}`);
            }}>Details</Button>
          {/* <Button variant="contained" color="secondary" onClick={handlePopoverClose}>X</Button> */}
        </Popover>
      </Grid>
      
      <Grid item xs={12} sm={2}>
        {/* <FormControl variant="filled" className={classes.formControl}>
          <InputLabel id="property-select-label">Change In</InputLabel>
          <Select
            labelId="property-select-label"
            id="property-select"
            value={property}
            onChange={({ target }) => setProperty(target.value)}
          >
            <MenuItem value="residentialChange">Residential</MenuItem>
            <MenuItem value="groceryChange">Grocery</MenuItem>
            <MenuItem value="parksChange">Parks</MenuItem>
            <MenuItem value="retailChange">Retail</MenuItem>
            <MenuItem value="transitChange">Transit</MenuItem>
            <MenuItem value="workplacesChange">Workplace</MenuItem>
          </Select>
        </FormControl> */}
        <Paper elivation={2} className={classes.legendPaper}>
          <FormControl component="fieldset">
            <FormLabel component="legend">Choose a Metric</FormLabel>
            <RadioGroup row aria-label="position" name="metric" defaultValue="retailChange">
              <FormControlLabel
                value="groceryChange"
                control={<Radio color="secondary"/>}
                label="Grocery"
              />
              <FormControlLabel
                value="parksChange"
                control={<Radio color="secondary"/>}
                label="Parks"
              />
              <FormControlLabel
                value="retailChange"
                control={<Radio color="secondary"/>}
                label="Retail"
              />
              <FormControlLabel
                value="transitChange"
                control={<Radio color="secondary"/>}
                label="Transit"
              />
              <FormControlLabel
                value="workplacesChange"
                control={<Radio color="secondary"/>}
                label="Workplace"
              />
            </RadioGroup>
          </FormControl>
        </Paper>
      </Grid>
      <Grid item xs={12}>
        {dates.length && <SliderStyled 
          value={dateIndex} 
          min={0} 
          max={dates.length - 1} 
          onChange={(_, newValue) => setDateIndex(newValue)} valueLabelDisplay="on" 
          valueLabelFormat={(index) => dates[index].slice(5)}
          marks={marks} />}
      </Grid>
    </Grid>
  </>  
  );
};

Map.propTypes = {
  countryCode: PropTypes.string,
  mapData: PropTypes.object.isRequired
};

export default Map;
