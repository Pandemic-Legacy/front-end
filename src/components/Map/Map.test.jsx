import React from 'react';
import { shallow } from 'enzyme';
import Map from './Map';
import { ThemeProvider } from '@material-ui/core';

describe('Map component', () => {
  it('renders Map', () => {
    const wrapper = shallow(
      <ThemeProvider>
        <Map />
      </ThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
