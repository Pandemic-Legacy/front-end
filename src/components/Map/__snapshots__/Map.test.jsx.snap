// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component renders Map 1`] = `
<ContextProvider>
  <Map />
</ContextProvider>
`;
